### ISSUE TITLE
!bad and the display_bad option broken

### ISSUE THREAD

**Issue by linas** [opencog/link-grammar#382]
The pull request #371 broke the !bad display option.  If !bad is set, then parsing should stop, instead of continuing to the next higher null count.

This capability is critical for debugging -- its a show-stopper.

**Comment by linas** [comment]
No its not #371, it is already broken in 5.3.2 ... wtf...

**Comment by linas** [comment]
Here's a way to test this:  add the following to the dictionary:
```
xxx: (Xd- & Xc+ & MX*a-);
yyy: (Xd- & Xc+ & MX*z-);
```
Compare the parses of
```
The lady, xxx, spoke sharply
The lady, yyy, spoke sharply
```

The first one should be blocked by post-processing, the second should parse fine.   Turning on !bad should show which post-processing rules are blocking the first parse (and should have zero null links, as a result).

**Comment by linas** [comment]
This is bizarre. The bug is a lot deeper than this. Parsing itself is giving inconsistent results.

**Comment by ampli** [comment]
Just note that #371 has no direct effect regarding the !bad flag, because link-parser has not been changed to do  "1-step" parsing. It still uses the 2-step algo as needed.

Regarding the broken case, I will look at it.

> This is bizarre. The bug is a lot deeper than this. Parsing itself is giving inconsistent results.
 Please provide an example, so I will also be able to investigate that.

**Comment by ampli** [comment]
>Here's a way to test this: add the following to the dictionary:
>[...]
> The lady, xxx, spoke sharply

It just has no parses. Should it have some?

With the unmodified dict:
>The lady, yyy, spoke sharply

LEFT-WALL the lady.n-f [,] [y] [,] spoke.v-d sharply

Note the **[y]**.
This is a very old bug: A spell guess changes **yyy** to **y**.
Now that the original word can be recovered from the wordgraph, I think that the result display should use the original word, or, may be better, a new display feature (turned on by !morphology=1) can indicate that **yyy** has been replaced by **y** due to a spell guess (this is more useful for parses in which the affected word is linked).

**Comment by linas** [comment]
turn off spell guessing. or change yyy to www or something that spell-guessing won't change.

The yyy senence should have parses. The xxx  sentence should have exactly the same parses, (maybe  bad-links printing needs to be turned on).

I've traced the bug to some power-pruning bug. -- the data coming into do_parse is already mangled.

**Comment by linas** [comment]
specifically, the number of disjuncts on word 0 (LEFT-WALL) differs for the two sentences, by the time that do_parse is called. The failing sentence has far too few disjuncts on it.

**Comment by linas** [comment]
Version 4.5.7 already has this bug ... wtf.

**Comment by linas** [comment]
If I comment out pp_and_power_prune, then both sentences start with the same number of disjuncts on the left-wall; however, this time, neither one parses.  How can that possibly happen?  Pruning is supposed to speed up parsing, but have no other side-effects...

**Comment by ampli** [comment]
The code that makes it parsable is in power_table_new(), with the comment:
```
   /* First, we initialize the word fields of the connectors, and
	  eliminate those disjuncts with illegal connectors */
```

If it is ifdef`ed 0, the the said sentence cannot be parsed.

**Comment by linas** [comment]
Oof-dah. Its pp_prune that is clobbering it!  Because pp_prune ignores the show_bad flag

**Comment by linas** [comment]
thanks Amir, I see .. I'm going to move that code around, now that I understand it...

**Comment by linas** [comment]
Fixed in pull req  #385
